app/src/main/cpp/skyline/services/nvdrv/core/nvmap.cpp

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: MIT OR MPL-2.0
+// Copyright © 2021 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#include "nvmap.h"
+
+namespace skyline::service::nvdrv::core {
+    NvMap::Handle::Handle(u64 size, Id id) : size(size), alignedSize(size), origSize(size), id(id) {}
+
+    PosixResult NvMap::Handle::Alloc(Flags pFlags, u32 pAlign, u8 pKind, u64 pAddress) {
+        std::scoped_lock lock(mutex);
+
+        // Handles cannot be allocated twice
+        if (allocated) [[unlikely]]
+            return PosixResult::NotPermitted;
+
+        flags = pFlags;
+        kind = pKind;
+        align = pAlign < PAGE_SIZE ? PAGE_SIZE : pAlign;
+
+        // This flag is only applicable for handles with an address passed
+        if (pAddress)
+            flags.keepUncachedAfterFree = false;
+        else
+            throw exception("Mapping nvmap handles without a cpu side address is unimplemented!");
+
+        size = util::AlignUp(size, PAGE_SIZE);
+        alignedSize = util::AlignUp(size, align);
+        address = pAddress;
+
+        // TODO: pin init
+
+        allocated = true;
+
+        return PosixResult::Success;
+    }
+
+    PosixResult NvMap::Handle::Duplicate(bool internalSession) {
+        // Unallocated handles cannot be duplicated as duplication requires memory accounting (in HOS)
+        if (!allocated) [[unlikely]]
+            return PosixResult::InvalidArgument;
+
+        std::scoped_lock lock(mutex);
+
+        // If we internally use FromId the duplication tracking of handles won't work accurately due to us not implementing
+        // per-process handle refs.
+        if (internalSession)
+            internalDupes++;
+        else
+            dupes++;
+
+        return PosixResult::Success;
+    }
+
+    NvMap::NvMap(const DeviceState &state) : state(state) {}
+
+    void NvMap::AddHandle(std::shared_ptr<Handle> handle) {
+        std::scoped_lock lock(handlesLock);
+
+        handles.emplace(handle->id, std::move(handle));
+    }
+
+    bool NvMap::TryRemoveHandle(const std::shared_ptr<Handle> &h) {
```

Comment:
We do use single letter variable names, see `i` in loops.
I think it makes sense here as it's consistent throughout the whole of nvdrv and it's what Nvidia uses too

Reply:
Also the cases where I use handle and id are different as they are technically different things in nvdrv despite being the same

Reply:
This is the only one where I feel somewhat strongly, I think the difference b/w ID/Handle would actually be accentuated if you didn't shorten it down to `h` as it took me a good bit to figure out that the ID was different from the handle and not just variable naming inconsistencies.

Reply:
Right... It's kinda Nvidia's fault here tbh just because of how they redid the ID system so it effectively matches the actual handle number all the time despite being different things. handleObj is just kinda excessively long for me...

Thinking about it would you rather me have places where handle is currently used be handleNum, and where the handle struct is used put just handle. Though imo I still prefer the original naming... Maybe handle for the integer and handleDesc for the struct? What do you think about that

Reply:
I somewhat proposed that with `handleId` but I think `handleNum` is also fine

Reply:
The thing is handleId sounds like ID which it isn't, I think I'll do handleDesc tbh, it works the best

Reply:
SGTM